Power Saving enhancements

psyke83 opened the pull request:
I figured out the issue with constant 2-3% CPU usage. When the audio driver is opened via SDL_OpenAudio, there will be a constant stream of failing ioctls from the ALSA driver, even when there's no active playback.

**strace -f -p $(pidof emulationstation)** will show this spammed constantly, even during idle & no background audio running: 

```
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = -1 EAGAIN (Resource temporarily unavailable)
4326  ioctl(4, SNDRV_PCM_IOCTL_WRITEI_FRAMES, 0x76f2edc0) = 0
4326  ioctl(4, SNDRV_PCM_IOCTL_SYNC_PTR, 0x2ee8f8) = 0
```
As I noted before, certain themes seem immune to the problem *until* after launching something. The reason is now clear:

- if a theme has a "scrollSound" element (such as Carbon), the ViewController::get()->preload() function that's called during early startup will cause the the audio device to open: https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/components/TextListComponent.h#L357. Anything that calls AudioManager::getInstance() or AudioManager's init() directly will trigger the bug. 
- if a theme has no "scrollSound" element, the preload function will never trigger a call to Sound::get(x), so the audio device will not be opened until ES returns from launching an item, here: https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/FileData.cpp#L202. This means that a theme with no scrollSound will not exhibit the extra CPU usage upon program start, but the CPU usage will occur after returning from a launched item due to the AudioManager instance being unnecessarily inited.

With this PS, the new behaviour is:

- ViewController::preload() will no longer trigger audio initialization on startup. Instead, onScroll will trigger audio init only when  playback is really needed.
- Audio is not explicitly re-initialized when returning from a launched item. Instead, I've made changes to ensure that when audio has been de-initialized, the re-initialization will trigger automatically for anything calling AudioManager::getInstance().
- Add an additional check in AudioManager::getInstance() so that audio init will not be triggered if navigation sounds are disabled. This will ensure that the audio device is never opened unnecessarily.
- The audio device will be closed during power saving mode (but only if no playback is detected as occurring).
- Force-disable audio in INSTANT PowerSaver mode. This is necessary to avoid lag introduced by constant deinit/reinit due to the 200ms delay between user input ending and audio deinit being triggered. 

The combined result is that idle CPU (with PS active) usage now goes down to 0.3%. Aside from CPU, keeping the audio device closed may possibly reduce power draw/heat, but that's only if the hardware audio codec powers down the chip (I've no idea about this, and don't have any means to test USB current draw).

hex007 commented:
Does this init & deinit SDL audio every time ES plays and stops playing sound? That would be horrible in case of scrolling a gamelist.

psyke83 commented:
No. The deinit is called here when ps_standby is true: https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/main.cpp#L314

While monitoring stract, I can see the deinit is called 5-10 seconds after scrolling one item. Also, I see no perceptible delay when the next scroll triggers reinit.

hex007 commented:
Have you tested it with INSTANT and ENHANCED modes?

psyke83 commented:
Enhanced will trigger deinit in about ~2 seconds. No perceptible delay.
Instant is a problem. I see a delay with this mode, so some changes will be necessary for the last commit.

psyke83 commented:
I'm curious: what's the purpose of the instant vs, say enhanced? Is there going to be a significant power savings between the two? Default and enhanced seem to work fine with my changes (but I'm on a Pi 3, so it may be a different story for older boards).

Can you suggest a better placement of the code to work with instant mode? I'm not certain if there's any point in me trying to fix it, since it will inevitably require adding a delay somewhere, which would make it  closer to the "enhanced" functionality.

hex007 commented:
Instant is used when the user cares not for animations. Specially useful for handheld systems where heating needs to be reduced. I had a Pizero handheld that used to get crazy hot (80%cpu) when Powersaver was not present. Enhanced was used when animations were required but nothing more. Default was needed as some users wanted the game description to scroll which didnt on I and E. Disabled is the behaviour when ES didnt have PS 

I will look into it this weekend.

psyke83 commented:
Thanks for the explanation. The lag is gone when instant is increased to 1000. I don't see an obvious solution to the problem, so I've added that change the PR. Of course, take your time to test and let me know if there's a better way to resolve the issue.

hex007 commented:
Changing the time out is not the way to solve your problem.

psyke83 commented:
No problem. Once you've tested and let me know, I can back out that commit at a later stage in place of other changes.

I think that the deinit/reinit is a worthy change in order to keep CPU usage to the absolute minimum. I have no idea if it affects actual power draw (which would depend on whether the Pi's audio codec can power down the hardware when the device is closed), but it's necessary to keep CPU usage at minimum when a theme is using navigation sounds. The other changes ensure that the audio device isn't opened unnecessarily in the first place, and don't depend on the PS deinit to be beneficial.

hex007 commented:
An alternative would be to disable all audio if PS is in Instant mode. This aligns with the intended reason for I mode.

psyke83 commented:
OK, but if you have navigation sounds disabled and/or use a theme with no scrollSound element present, because of my first two commits, AudioManager will never call init unnecessarily. This means there'll be no lag problem in any PS mode. SDL_GetAudioStatus() will report SDL_AUDIO_STOPPED when the device is not opened or in an unknown state, never SDL_AUDIO_PAUSED, so the deinit will not be triggered unnecessarily, either.

jrassa commented on es-app/src/components/TextListComponent.h:
```diff
@@ -76,13 +77,11 @@ class TextListComponent : public IList<TextListData, T>
 
 protected:
-	virtual void onScroll(int amt) { if(mScrollSound) mScrollSound->play(); }
+	virtual void onScroll(int amt) { if(!mScrollSound.empty() && Settings::getInstance()->getBool("EnableSounds")) { Sound::get(mScrollSound)->play(); } }
```
As an alternative to having to check the setting everywhere that we may want to play a sound.  Could we have something along the lines of `Sound::play(mScrollSound)` that handles it?

psyke83 replied:
There is already a check here: https://github.com/RetroPie/EmulationStation/blob/master/es-core/src/Sound.cpp#L114

However, Sound::get() itself will trigger AudioManager::getInstance(), which will open the audio device *before* reaching the Sound::play() check (even if EnableSounds is false). Adding the additional check here avoids unnecessarily opening the audio device in this circumstance.

jrassa replied:
I understand that, but ideally the TextListComponent shouldn't even be aware that there is a setting to enable/disable sounds.    To be more clear, I am suggesting adding something along the lines of the following to Sound.cpp: 

```cpp
void Sound::play(const std::string& path)
{
     if(!mScrollSound.empty() && Settings::getInstance()->getBool("EnableSounds")) 
    { 
        Sound::get(mScrollSound)->play(); 
    }
}
```

Also, shouldn't these need to be adjusted as well?

https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/views/gamelist/IGameListView.cpp#L16

https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/views/gamelist/ISimpleGameListView.cpp

psyke83 replied:
Thanks for the feedback - I've updated the patchset.

TextListComponent no longer checks the bool. Instead a check is added to AudioManager::getInstance(), which means that it will also catch the extra cases you identified.

psyke83 commented:
@hex007,

I updated the PS due to @jrassa's comments, but I also changed the 1sec delay, as it was causing the game info to start scrolling for a split second. Reducing to 800ms solves the unwanted scrolling & lag is still not noticeable. Will be waiting for a better solution after you've tested, though the current change to the delay seems OK to me.

psyke83 commented:
I've updated the first post to incorporate the information I posted on the old bug report, just so everybody else is up to speed.

@Hex007, a thought re: PowerSaver INSTANT. If you don't like the behaviour with the 800ms delay, we could add a corner case so that deinit is only triggered when the screensaver is active, but only when PS is in INSTANT mode. This would still be a compromise, as ES would have the 3% CPU issue in INSTANT mode until the screensaver kicked in (IFF sounds are enabled; otherwise, the other commits fix the problem).

hex007 commented:
@psyke83 I ran your branch and followed the instructions. I am still getting the error.
This is with PS disabled.

```
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid 11610] recvmsg(4, {msg_namelen=0}, 0) = -1 EAGAIN (Resource temporarily unavailable)
```

psyke83 commented:
@hex007 

That's the expected behaviour, partially because it's necessary to have a reliable way to ensure a delay between user input stopping and invoking the deinit.

I added the deinit to PowerSaver::getTimeout(), which is only invoked when ps_standby is set to true here: https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/main.cpp#L314

So, audio deinit will occur in any PS state except disabled. This seems fairly consistent, given that deinit is a power saving feature.

If your active theme has no scroll/launch/back sound, or even if it does, but you've set navigation sounds to disabled, audio will never be inited in the first place.

psyke83 commented:
If you want to avail of audio deinit even when PS is disabled, an additional check can be placed here: https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/main.cpp#L357

This will deinit audio when the screensaver kicks in. The call in PowerSaver::getTimeout() could be left, but changed to exclude processing in instant mode so that the addition to main.cpp would cover the disabled and instant cases.

jrassa commented:
> @hex007, a thought re: PowerSaver INSTANT. If you don't like the behaviour with the 800ms delay, we could add a corner case so that deinit is only triggered when the screensaver is active, but only when PS is in INSTANT mode. This would still be a compromise, as ES would have the 3% CPU issue in INSTANT mode until the screensaver kicked in (IFF sounds are enabled; otherwise, the other commits fix the problem).

@psyke83 I would recommend disabling audio automatically when PowerSaver is set to INSTANT.  This is already done for animations. 

See here:
https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/guis/GuiMenu.cpp#L267

psyke83 commented:
@jrassa done. Instant is back to the old timeout value and sounds will be force-disabled.

What about audio deinit when PS is disabled? Is it necessary to implement it?

pjft commented:
In general, or in any specific case outside of instant?

If I'm following correctly, you mentioned the audio deinit was meant to
reduce CPU consumption drastically, so what are you proposing/asking?
On Fri, 15 Sep 2017 at 04:57 psyke83 <notifications@github.com> wrote:

> @jrassa <https://github.com/jrassa> done. Instant is back to the old
> timeout value and sounds will be force-disabled.
>
> What about audio deinit when PS is disabled? Is it necessary to implement
> it?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/242#issuecomment-329673807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7b64Z1fSnlA-sTjX9Yt9IkeAUmE_ks5sifW-gaJpZM4PW76c>
> .
>

hex007 commented:
@pjft Disabling audio further reduces cpu consumption by ES from 3% to 0.3% on a PiZ so cant call it drastic. 

@psyke83 When PS is set to DISABLED (please call it like that to avoid confusion), irrespective of audio state, deinit should not be called.

Can you find a state where the error messages are not present? Implementing that would be the correct direction. Try writing a clean code example that uses SDL audio for testing this concept. Also please explain what is the meaning of the error you reported. I am having trouble understanding what is the error.

I searched around and `EAGAIN (Resource temporarily unavailable)` is actually not a bug. It has to do with reading and writing of socket streams. 

PS does cause a warning which prints something like "XXms lag". This was harmless and hence ignored.

pjft commented:
I'd rather have a current summary of where we are at currently with each power state, and then make a decision.

While I understand the difficulties with Instant, and I'll leave that decision at your discretion, if there's a benefit for other power saving modes we should explore it. If I'm running a Pi 24/7 with ES, I'd rather have it a 0.3% all the time rather than 3% - it is a 90% reduction, even though it's already at quite a low power consumption level.

If nothing more, I'd at least support deinitting and re-initting on launch/exit of the Black and Dim screensavers.

psyke83 commented:
It's a non fatal error, I agree. I tested a basic SDL implementation of aplay (.wav player) and the issue also occurs. The error can be avoided if you change audio format  to signed 32bit. ES CPU then fluctuates between 2% and 0.3% - better, but the constant stream of ioctls from the kernel driver causes the very minor spike.

Whether there's a bug in the driver is irrelevant. The ioctl spam simply helped identify issues with unnecessary audio init in the program.

psyke83 commented:
@hex007 So you're now happy with the behaviour of this patchset for the INSTANT and DISABLED cases?

@pjft 

Here's a summary:

General: 
* audio init is no longer called unnecessarily on startup (during preload), or when returning from launching an item.
* Audio init will now first occur only at the first moment ES plays a sound. If a theme has no sounds, or navigation sounds are set to disabled, audio init will never occur.
* Once audio has been initialized, the 3% CPU "bug" will be triggered.

PowerSaver changes:
* When the PowerSaver::getTimeout() function is called, audio will be deinitialized if the SDL subsystem reports the audio device as paused (i.e., the device is open but nothing is being played). This is triggered for all PowerSaver modes except DISABLED.
* PowerSaver's INSTANT mode now force-disables sound. This is by @hex007's request. Without this change, audio deinit/reinit would occur far too frequently, leading to noticeable lag when scrolling items on a theme with scroll sounds.
* If PowerSaver is set to DISABLED, the 3% audio bug will persist when navigation sounds are used. This is why I was asking if we should add an additional case, but I think that leaving this behaviour intact is fairly consistent.

pjft commented:
Thank you.

So, to confirm:
- Audio De-init will occur if whenever getTimeout() gets called (which, I
believe, is when the PS timeout gets checked or woken up) and no audio is
playing.
- Audio re-init will only occur when a sound plays.
- Instant has no sound. Disabled works as previously.

Looks good to me. Thank you for your work!

psyke83 commented:
@pjft 

Pretty much, with the caveat that PS = DISABLED will no longer trigger the 3% "bug" if ES never plays sound during an active session. If you temporarily enable navigation sounds, trigger a scroll sound, and disable, the 3% CPU bug will trigger. However, launching an item and returning will clear the issue due to the explicit audio deinit (intended to avoid conflicts with launched applications).

joolswills commented:
Is this ready to go ? Has anyone had a chance to test it ?

hex007 commented:
@psyke83 Is still working on this. Once things are updated @pjft and I will test it. If it is ready to go then we shall inform you.

psyke83 commented:
It's been ready to test since a month ago, yes. I rebased the patches recently, but otherwise no changes. I've seen no problems, but I haven't really tested background audio/videos yet.

hex007 commented:
I apologise for taking so long to get to this. I tried this patch for a week and did not notice any performance advantage (without using top/strace) 

That said it adds considerable complexity to the code to maintain it further. 

Considering that we might shift to some other means of audio than SDL audio, which blocks other applications from using audio device, I would not recommend merging this at present. 

I would like to work on incorporating/ testing said Audio alternative from ground up so this becomes easier. I am currently reading up on such methods.

@pjft thoughts?

pjft commented:
> Considering that we might shift to some other means of audio than SDL audio, which blocks other applications from using audio device, I would not recommend merging this at present.

SDL is a well known cross-platform library to allow for this exact purpose. What are the plans to move away from SDL, and what's the driving force behind that? 

From looking at it, the code itself doesn't strike me as overly complex in terms of changes, and if it improves things, I'm certainly not one to oppose. 

Are there areas of the code we feel could be simplified in order to ensure better maintainability in the future?

hex007 commented:
>From looking at it, the code itself doesn't strike me as overly complex in terms of changes, and if it improves things, I'm certainly not one to oppose.

In that case @psyke83 please squash the commits. @pjft if you agree to the merge, please let buzz know.

>Are there areas of the code we feel could be simplified in order to ensure better maintainability in the future?

This can be worked on later.

psyke83 commented:
I'd prefer this not be merged if there's no agreement on its utility. The benefit is extremely minuscule, but the idea was to improve potential battery life due to the audio codec being constantly powered for no reason.

Also, I'm not so sure it adds much complexity; the bulk of the patch is blacklisting the INSTANT powersave case, which was by request on previous reviews.

If you're not comfortable with the deinit on pause, then I can remove these commits:
* PowerSaver: close audio during PS when possible
* PowerSaver: force-disable sounds for INSTANT profile

This will leave the commits that improve the behaviour of audio init; i.e. audio will only init at the first moment an actual sound needs to play, and will not explicitly re-init after returning from a launched item.

The downside will be that the 3% CPU usage will occur after sound playback if navigation sounds are turned on, but the "bug" will clear after returning from a launched item. Additionally, the issue will never trigger if navigation sounds are disabled by the user.

I'm curious why you're thinking of moving away from SDL2, and what API you had in mind to replace it (direct ALSA?)

hex007 commented:
@psyke83 This can be merged in dev branch for wider testing. This is my opinion. The feature has been checked by very few devs. Just to e safe I want more testing. We were discussing on having a dev branch and a production branch to keep things stable. There were minor hiccups with PS in the past and I dont want to repeat that experience thats all. This feature is a good point to start dev branch hence the suggestion. I trust @pjft 's opinion and if he says its ready then I will support it too. 

> I'm curious why you're thinking of moving away from SDL2, and what API you had in mind to replace it (direct ALSA?)

There were some talks going on, when I was last active on forums, about Audio problems related to SDL audio. I just assumed they might have developed somehow and there might be an interest in that direction. 

Currently SDL hogs the audio sink so no other application can use audio if SDL audio is initialized, which is not convenient.

joolswills commented:
The master branch is our current "dev" branch.

I've not seen any discussion about moving away from SDL for audio, and there are no plans to move.

hex007 commented:
@joolswills Glad to know. Are we switching to dev and production branch model? I found this article really helpful. http://nvie.com/posts/a-successful-git-branching-model/

joolswills commented:
@hex007 we already did, or at least retropie builds off the stable branch now which I merge when I'm happy with the status of master.

pjft commented:
Sorry for the delay and limited availability in the recent past. I tested this, with and without VLC. It seems to work as intended, I am supportive or merging it. 

Thanks for this.
